src/main/java/com/streamr/client/subs/CombinedSubscription.java

```diff
@@ -0,0 +1,93 @@
+package com.streamr.client.subs;
+
+import com.streamr.client.MessageHandler;
+import com.streamr.client.exceptions.GapDetectedException;
+import com.streamr.client.exceptions.UnableToDecryptException;
+import com.streamr.client.exceptions.UnsupportedMessageException;
+import com.streamr.client.options.ResendOption;
+import com.streamr.client.protocol.message_layer.StreamMessage;
+import com.streamr.client.utils.OrderedMsgChain;
+
+import java.util.ArrayDeque;
+import java.util.Map;
+
+public class CombinedSubscription extends Subscription {
+    private BasicSubscription sub;
+    private final ArrayDeque<StreamMessage> queue = new ArrayDeque<>();
+    public CombinedSubscription(String streamId, int partition, MessageHandler handler, ResendOption resendOption, Map<String, String> groupKeysHex,
+                                long propagationTimeout, long resendTimeout) {
+        super(streamId, partition, handler, groupKeysHex, propagationTimeout, resendTimeout);
+        MessageHandler wrapperHandler = new MessageHandler() {
+            @Override
+            public void onMessage(Subscription sub, StreamMessage message) {
+                handler.onMessage(sub, message);
+            }
+            @Override
+            public void done(Subscription s) {
+                handler.done(s);
+                RealTimeSubscription realTime = new RealTimeSubscription(streamId, partition, handler, groupKeysHex, propagationTimeout, resendTimeout);
+                realTime.setGapHandler(sub.getGapHandler());
+                realTime.setLastMessageRefs(sub.getChains());
+                while(!queue.isEmpty()) {
+                    StreamMessage msg = queue.poll();
+                    realTime.handleRealTimeMessage(msg);
+                }
+                sub = realTime;
+            }
+        };
+        sub = new HistoricalSubscription(streamId, partition, wrapperHandler, resendOption, groupKeysHex, propagationTimeout, resendTimeout, msg -> {
+            queue.push(msg);
+            return null;
```

Comment:
Btw to avoid having to write these `return null` statements you could use `Consumer<T>` instead of `Function<T, Void>`. See https://docs.oracle.com/javase/8/docs/api/java/util/function/Consumer.html

Reply:
Oh nice! :)